src/node/txreconciliation.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <node/txreconciliation.h>
+
+#include <unordered_map>
+
+namespace {
+
+/** Current protocol version */
+constexpr uint32_t RECON_VERSION = 1;
+/** Static salt component used to compute short txids for sketch construction, see BIP-330. */
+const std::string RECON_STATIC_SALT = "Tx Relay Salting";
+
+/**
+ * Salt (specified by BIP-330) constructed from contributions from both peers. It is used
+ * to compute transaction short IDs, which are then used to construct a sketch representing a set
+ * of transactions we want to announce to the peer.
+ */
+static uint256 ComputeSalt(uint64_t local_salt, uint64_t remote_salt)
+{
+    // According to BIP-330, salts should be combined in ascending order.
+    uint64_t salt1 = local_salt, salt2 = remote_salt;
+    if (salt1 > salt2) std::swap(salt1, salt2);
+
+    static const auto RECON_SALT_HASHER = TaggedHash(RECON_STATIC_SALT);
+    return (CHashWriter(RECON_SALT_HASHER) << salt1 << salt2).GetSHA256();
+}
+
+/**
+ * Used to keep track of the ongoing reconciliations, the transactions we want to announce to the
+ * peer when next transaction reconciliation happens, and also all parameters required to perform
+ * reconciliations.
+ */
+class ReconciliationState {
+
+    public:
+
+    /**
+     * Reconciliation involves exchanging sketches, which efficiently represent transactions each
+     * peer wants to announce. Sketches are computed over transaction short IDs.
+     * These values are used to salt short IDs.
+     * TODO: These are currently public to overcome compiler warning (-Wunused-private-field).
+     *       They should be made private once they are used in one of the following commits.
+     */
+    const uint64_t m_k0, m_k1;
+
+    /**
+     * Reconciliation protocol assumes using one role consistently: either a reconciliation
+     * initiator (requesting sketches), or responder (sending sketches). This defines our role.
+     * */
+    const bool m_we_initiate;
+
+    ReconciliationState(uint64_t k0, uint64_t k1, bool we_initiate) :
+        m_k0(k0), m_k1(k1), m_we_initiate(we_initiate) {}
+};
+
+} // namespace
+
+/** Actual implementation for TxReconciliationTracker's data structure. */
+class TxReconciliationTracker::Impl {
+
+    mutable Mutex m_mutex;
+
+    /**
+     * Per-peer salt is used to compute transaction short IDs, which will be later used to
+     * construct reconciliation sketches.
+     * Salt is generated randomly per-peer to prevent:
+     * - linking of network nodes belonging to the same physical node
+     * - halting of relay of particular transactions due to short ID collisions (DoS)
+     */
+    std::unordered_map<NodeId, uint64_t> m_local_salts GUARDED_BY(m_mutex);
```

Comment:
1. A new member would sit there with no use most of the `ReconciliationState` lifetime
2. `ReconciliationState` would be initialized but not ready to use for reconciliations (while there is no full salt)

I thought the latter might be particularly confusing in terms of understanding the protocol flow

Reply:
I started integrating `m_local_salt` into `ReconciliationState`, and realized that I'm not sure what's the best way to check the state of the peer (pre-registered or registered)?
1. A new `m_registrered` flag in `ReconciliationState`
2. Make `m_local_salt` in ReconciliationState optional, and if it's nullptr -- means it's registered already (more implicit approach)
3. Something even crazier is having `std::variant <std::pair<NodeId, uint64_t>, ReconciliationState> m_states`, so that the value could be either a node<->salt mapping (pre-registered) and `ReconciliationState` (registered)

@glozow have an opinion?

Reply:
I agree with having the salt inside of `ReconciliationState`, and it seems like (2) would be most idiomatic - `std::optional<ReconciliationState>` where nullopt means not registered yet.